Demand and DSR documentation (#295)
open-energy-transition/gb-dispatch-model: doc/gb-model/demand_and_dsr.rst

..
  SPDX-FileCopyrightText: gb-dispatch-model contributors
  SPDX-License-Identifier: CC-BY-4.0

.. _demand_and_dsr:

##########################

> Demand and DSR

Reviewer: ```suggestion
Baseline demand & demand-side response (DSR)
```

##########################

Overview
========

> The GB dispatch model incorporates demand-side response (DSR) capabilities across multiple demand sectors to provide flexibility in the electricity system. DSR allows for the shifting of electricity consumption from peak to off-peak periods, helping to balance supply and demand while reducing the need for additional generation capacity.

Reviewer: ```suggestion
The GB dispatch model incorporates demand-side response (DSR) capabilities across multiple demand sectors to provide flexibility in the electricity system. 
DSR allows for the shifting of electricity consumption from peak to off-peak periods, helping to balance supply and demand while reducing the need for additional generation capacity.
```

.. graphviz::

  digraph Flow {
      rankdir=LR;   // Left to Right

      node [shape=box, style=filled];

      AC_bus [label="AC Bus", fillcolor="#B3D9FF", shape=ellipse, width=2, height=1.5, fixedsize=true];
      Sector_bus [label="Sector Bus", fillcolor="#FFD1DC", shape=ellipse, width=1.8, height=1.2, fixedsize=true];
      Sector_DSR_bus [label="Sector DSR Bus", fillcolor="#FFD1DC", shape=ellipse, width=1.8, height=1.2, fixedsize=true];
      heat_load [label="Sector Load", fillcolor="#FFDAB9"];
      sector_dsr_store [label="Sector DSR Store", fillcolor="#FFFACD"];

      AC_bus -> Sector_bus [label="unmanaged load link"];
      Sector_bus -> heat_load;
      Sector_bus -> Sector_DSR_bus [label="charge"];
      Sector_DSR_bus -> Sector_bus [label="discharge"];
      Sector_DSR_bus -> sector_dsr_store [dir=both];
  }

"Sector" refers to the specific demand sector being modeled (e.g., residential, I&C, EV, heat or hydrogen).
Whilst the general structure is consistent across demand types, some extra components are added for specific sectors which can be found in the corresponding documentation sections (e.g., EV charging includes additional V2G storage components).

> 

Reviewer: Add ref here to the other subsystem pages for more info.

> 

Reviewer: ```suggestion

```

Data Sources
============

Great Britain data
------------------

> Demand and flexibility data are sourced from the Future Energy Scenario FES 2024 workbooks. The model processes FES data to extract:

Reviewer: ```suggestion
Demand and flexibility data are sourced from the Future Energy Scenario FES 2024 workbooks.
The model processes FES data to extract:
```

- Annual electricity demand by sector and region
- DSR capacity and flexibility potential
- Technology-specific demand characteristics

The table below summarizes the specific sheets extracted from the FES workbook for each of the demand sectors:

.. _demand_and_dsr_data_sources:
.. list-table:: FES Data Sources for Demand and DSR
   :header-rows: 1

   * - Demand Sector
     - FES Workbook Sheet
   * - Baseline Electricity (Residential)
     - Building Block Data (BB1)
   * - Baseline Electricity (I&C)
     - Building Block Data (BB1)
   * - Residential Heat
     - Building Block Data (BB1), Gas Demand Summary (ED3) and Flexibility (FLX1)
   * - I&C Heat
     - Building Block Data (BB1), Gas Demand Summary (ED3) and Flexibility (FLX1)
   * - EV Charging

>      - Building Block Data (BB1), FES ED5(EV Demand) and Flexibility (FLX1)

Reviewer: ```suggestion
     - Building Block Data (BB1), Road Transport Summary (ED5) and Flexibility (FLX1)
```

  This is specific for EV DSR and is linked to vehicle availability patterns.

Data Processing Workflow
========================

The demand and DSR implementation follows a multi-step workflow:

1. **Demand Table Creation** - Extract annual demand data from FES workbook by technology and region
2. **Flexibility Table Creation** - Process DSR capacity data from FES flexibility sheets
3. **Regional Distribution** - Distribute national data to regional PyPSA buses
4. **Demand Profile Scaling** - Apply temporal profiles to annual demands
5. **Network Composition** - Add DSR components to the PyPSA network

Data Processing Pipeline
=========================

The demand and DSR workflow is implemented through Snakemake rules in `rules/gb-model/demand_and_dsr.smk` and Python scripts in the folder `scripts/gb_model/demand_and_dsr`.

The rulegraph for the demand and DSR workflow is illustrated below:
.. image:: img/demand_and_dsr.svg
   :align: center

.. note::
  The graph above was generated using::

      pixi run filtered_rulegraph \
      "resources/GB/gb-model/HT/baseline_electricity_demand/2030.csv" \
      "resources/GB/gb-model/HT/additional_demand/2030.csv" \

>       " resources/GB/gb-model/HT/residential_dsr.csv" \

Reviewer: ```suggestion
      " resources/GB/gb-model/HT/regional_iandc_dsr_inc_eur.csv" \
      " resources/GB/gb-model/HT/regional_residential_dsr_inc_eur.csv" \
```

This aligns with the outputs of the rule that we then use in `rules/gb-model.smk::compose_network`. Actual SVG will need updating, too.

..
  SPDX-FileCopyrightText: gb-dispatch-model contributors
  SPDX-License-Identifier: CC-BY-4.0

.. _demand_and_dsr:

##########################
Demand and DSR
##########################

Overview
========

The GB dispatch model incorporates demand-side response (DSR) capabilities across multiple demand sectors to provide flexibility in the electricity system. DSR allows for the shifting of electricity consumption from peak to off-peak periods, helping to balance supply and demand while reducing the need for additional generation capacity.

Demand Sectors
---------------

The model includes the following demand sectors with DSR capabilities:

1. **Baseline Electricity (Residential)** - Residential electricity demand excluding heat pumps
2. **Baseline Electricity (I&C)** - Industrial and Commercial electricity demand
3. **Residential Heat** - Electricity demand for residential heating systems
4. **I&C Heat** - Electricity demand for industrial and commercial heating systems
5. **EV Charging** - Electric vehicle charging demand

> 6. **Hydrogen Electrolyser Demand** - Electricity demand for hydrogen production via electrolysis

Reviewer: There's no hydrogen DSR but rather hydrogen storage, so this shouldn't be here.

Great Britain data
------------------
Demand and flexibility data are sourced from the Future Energy Scenario FES 2024 workbooks. The model processes FES data to extract:

- Annual electricity demand by sector and region
- DSR capacity and flexibility potential
- Technology-specific demand characteristics

The table below summarizes the specific sheets extracted from the FES workbook for each of the demand sectors:

.. _demand_and_dsr_data_sources:
.. list-table:: FES Data Sources for Demand and DSR
   :header-rows: 1

   * - Demand Sector
     - FES Workbook Sheet
   * - Baseline Electricity (Residential)
     - Building Block Data (BB1)
   * - Baseline Electricity (I&C)
     - Building Block Data (BB1)
   * - Residential Heat
     - Building Block Data (BB1), Gas Demand Summary (ED3) and Flexibility (FLX1)
   * - I&C Heat
     - Building Block Data (BB1), Gas Demand Summary (ED3) and Flexibility (FLX1)
   * - EV Charging
     - Building Block Data (BB1), FES ED5(EV Demand) and Flexibility (FLX1)
   * - Hydrogen Electrolyser Demand
     - Building Block Data (BB1), WS1 (Whole System)
   * - Additional demand
     - Electricity Demand Summary (ED1)

European data
---------------

> For European neighbour countries included in the model, the demand data is sourced from the ENTSO-E energy balance data (via ``energy_totals.csv``). Specific data requirements are explained in detail in the corresponding sections of the documentation (e.g., :ref:`heat_system` uses ES2 from the FES workbook for European demand assumptions).

Reviewer: Suitably short, but the annual demand is estimated from the total country demand in ``ES2``, split into end-use sectors using ``energy_totals.csv``. This isn't from ENTSO-E but rather Eurostat/JRC-IDEES (we need to edit that in EV page, too). No need to give the source, though since we can say we get it from PyPSA-Eur and point to the relevant section of the PyPSA-Eur docs that talks about it.

Configuration
=============

DSR is configured in the model configuration files under the ``fes.gb.flexibility`` section:

> .. code-block:: yaml

Reviewer: Use the `literalinclude` directive as in EV page

System Components
==================

DSR is implemented using PyPSA's Bus, Store, Link and Load components to model energy storage and power flow capabilities. For each demand sector with DSR, the model creates:

- **DSR Storage Bus** - A dedicated bus for storing shifted energy
- **Shift Link** - Allows energy to be moved from the main demand bus to storage during off-peak periods
- **Reverse Link** - Allows energy to be returned from storage to the main demand bus during peak periods
- **Store Component** - Represents the energy storage capacity with time-dependent constraints
- **Load Component** - Represents the total demand that must be met, including both the baseline demand and any additional demand from DSR.

The representation of each demand sector is illustrated in the :ref:`_model_representation` section of the documentation, which provides detailed descriptions of the components and their interactions.

Implementation Notes
====================

Model Representation
--------------------

.. _model_representation:

The DSR implementation for each demand type (residential, I&C, EV and heat) follows this structure:

> .. graphviz::

Reviewer: Move up to overview

      AC_bus [label="AC Bus", fillcolor="#B3D9FF", shape=ellipse, width=2, height=1.5, fixedsize=true];
      Sector_bus [label="Sector Bus", fillcolor="#FFD1DC", shape=ellipse, width=1.8, height=1.2, fixedsize=true];
      Sector_DSR_bus [label="Sector DSR Bus", fillcolor="#FFD1DC", shape=ellipse, width=1.8, height=1.2, fixedsize=true];
      heat_load [label="Sector Load", fillcolor="#FFDAB9"];
      sector_dsr_store [label="Sector DSR Store", fillcolor="#FFFACD"];

      AC_bus -> Sector_bus [label="unmanaged load link"];
      Sector_bus -> heat_load;
      Sector_bus -> Sector_DSR_bus [label="charge"];
      Sector_DSR_bus -> Sector_bus [label="discharge"];
      Sector_DSR_bus -> sector_dsr_store [dir=both];
  }

"Sector" refers to the specific demand sector being modeled (e.g., residential, I&C, EV, heat or hydrogen).
Whilst the general structure is consistent across demand types, some extra components are added for specific sectors which can be found in the corresponding documentation sections (e.g., EV charging includes additional V2G storage components).

DSR Parameters
--------------

DSR operation is controlled by several key parameters:

**DSR Hours**

>   Time windows during which DSR can operate. Configured per sector:

Reviewer: ```suggestion
  Time windows during which DSR can operate. 
  These time windows are not published by NESO, so we have assumed values per sector as follows:
```

  - Residential baseline electricity: 17:00-20:00 (5pm-8pm)
  - I&C baseline electricity: 17:00-20:00 (5pm-8pm)
  - Residential heat: 00:00-22:00 (all day)
  - I&C heat: 00:00-22:00 (all day)
  - EV charging: 08:00-06:00 (8am-6am next day)

**Storage Capacity**
  Calculated as DSR power capacity × duration hours. Represents the maximum energy that can be shifted.

**Availability Profiles**

>   Time-dependent constraints on when DSR can operate. This is specific for EV DSR and is linked to vehicle availability patterns.

Reviewer: one line per sentence

European Integration
--------------------

For European neighbour countries included in the model, demand and DSR data are synthesized by:

1. Using GB data as a reference for technology mixes
2. Scaling based on relative annual demand totals from energy balance data
3. Applying appropriate time zone shifts for DSR operation windows

> This ensures consistent representation of cross-border demand flexibility effects.

Reviewer: Not sure this is needed

Author: removed it

Data Processing Workflow
========================

The demand and DSR implementation follows a multi-step workflow:

1. **Demand Table Creation** - Extract annual demand data from FES workbook by technology and region
2. **Flexibility Table Creation** - Process DSR capacity data from FES flexibility sheets
3. **Regional Distribution** - Distribute national data to regional PyPSA buses
4. **Demand Profile Scaling** - Apply temporal profiles to annual demands
5. **Network Composition** - Add DSR components to the PyPSA network

Data Processing Pipeline
=========================

The demand and DSR workflow is implemented through Snakemake rules in `rules/gb-model/demand_and_dsr.smk` and Python scripts in the folder `scripts/gb_model/demand_and_dsr`.

The rulegraph for the demand and DSR workflow is illustrated below, showing the key processing steps and their dependencies:

.. image:: img/demand_and_dsr.svg
   :align: center

> 

Reviewer: Give note on how image was generated (see EV page)
